# Import

In [2]:
import os
import pandas as pd

# Function: File

In [3]:
def get_files(dir_):
    return [f for f in os.listdir(dir_) if '.csv' in f]


def validate_file(file):
    with open(file) as f:
        text = f.read()
    return not (('fail' in text) and ('error' in text))


def get_valid_files(dir_):
    return [f for f in get_files(dir_) if validate_file(dir_ + f)]


def get_hidden_region_from_file(file):
    return file.split('-')[1]


def get_downloaded_regions(dir_):
    return [get_hidden_region_from_file(file) for file in get_valid_files(dir_) if '.csv' in file]


dir_item = '/home/jake/Documents/News_Item/Browser/'
dir_arch = dir_item + 'archive/'
dir_arch2 = dir_item + 'archive2/'

files = get_valid_files(dir_item)
print(files[:2])
len(files)

['browser-TH-monthly-201801-201901.csv', 'browser-AD-monthly-201801-201901.csv']


257

# Const

In [4]:
countries = [('ww', 'Worldwide'), ('af', 'Africa'), ('an', 'Antarctica'), ('as', 'Asia'), ('eu', 'Europe'), ('na', 'North America'), ('oc', 'Oceania'), ('sa', 'South America'), ('AF', 'Afghanistan'), ('AX', 'Åland Islands'), ('AL', 'Albania'), ('DZ', 'Algeria'), ('AS', 'American Samoa'), ('AD', 'Andorra'), ('AO', 'Angola'), ('AI', 'Anguilla'), ('AQ', 'Antarctica'), ('AG', 'Antigua and Barbuda'), ('AR', 'Argentina'), ('AM', 'Armenia'), ('AW', 'Aruba'), ('AU', 'Australia'), ('AT', 'Austria'), ('AZ', 'Azerbaijan'), ('BS', 'Bahamas'), ('BH', 'Bahrain'), ('BD', 'Bangladesh'), ('BB', 'Barbados'), ('BY', 'Belarus'), ('BE', 'Belgium'), ('BZ', 'Belize'), ('BJ', 'Benin'), ('BM', 'Bermuda'), ('BT', 'Bhutan'), ('BO', 'Bolivia, Plurinational State of'), ('BQ', 'Bonaire, Sint Eustatius and Saba'), ('BA', 'Bosnia and Herzegovina'), ('BW', 'Botswana'), ('BV', 'Bouvet Island'), ('BR', 'Brazil'), ('IO', 'British Indian Ocean Territory'), ('BN', 'Brunei Darussalam'), ('BG', 'Bulgaria'), ('BF', 'Burkina Faso'), ('BI', 'Burundi'), ('KH', 'Cambodia'), ('CM', 'Cameroon'), ('CA', 'Canada'), ('CV', 'Cape Verde'), ('KY', 'Cayman Islands'), ('CF', 'Central African Republic'), ('TD', 'Chad'), ('CL', 'Chile'), ('CN', 'China'), ('CX', 'Christmas Island'), ('CC', 'Cocos (Keeling) Islands'), ('CO', 'Colombia'), ('KM', 'Comoros'), ('CG', 'Congo'), ('CD', 'Congo, the Democratic Republic of the'), ('CK', 'Cook Islands'), ('CR', 'Costa Rica'), ('CI', "Côte d'Ivoire"), ('HR', 'Croatia'), ('CU', 'Cuba'), ('CW', 'Curaçao'), ('CY', 'Cyprus'), ('CZ', 'Czech Republic'), ('DK', 'Denmark'), ('DJ', 'Djibouti'), ('DM', 'Dominica'), ('DO', 'Dominican Republic'), ('EC', 'Ecuador'), ('EG', 'Egypt'), ('SV', 'El Salvador'), ('GQ', 'Equatorial Guinea'), ('ER', 'Eritrea'), ('EE', 'Estonia'), ('ET', 'Ethiopia'), ('FK', 'Falkland Islands (Malvinas)'), ('FO', 'Faroe Islands'), ('FJ', 'Fiji'), ('FI', 'Finland'), ('FR', 'France'), ('GF', 'French Guiana'), ('PF', 'French Polynesia'), ('TF', 'French Southern Territories'), ('GA', 'Gabon'), ('GM', 'Gambia'), ('GE', 'Georgia'), ('DE', 'Germany'), ('GH', 'Ghana'), ('GI', 'Gibraltar'), ('GR', 'Greece'), ('GL', 'Greenland'), ('GD', 'Grenada'), ('GP', 'Guadeloupe'), ('GU', 'Guam'), ('GT', 'Guatemala'), ('GG', 'Guernsey'), ('GN', 'Guinea'), ('GW', 'Guinea-Bissau'), ('GY', 'Guyana'), ('HT', 'Haiti'), ('HM', 'Heard Island and McDonald Islands'), ('VA', 'Holy See (Vatican City State)'), ('HN', 'Honduras'), ('HK', 'Hong Kong'), ('HU', 'Hungary'), ('IS', 'Iceland'), ('IN', 'India'), ('ID', 'Indonesia'), ('IR', 'Iran, Islamic Republic of'), ('IQ', 'Iraq'), ('IE', 'Ireland'), ('IM', 'Isle of Man'), ('IL', 'Israel'), ('IT', 'Italy'), ('JM', 'Jamaica'), ('JP', 'Japan'), ('JE', 'Jersey'), ('JO', 'Jordan'), ('KZ', 'Kazakhstan'), ('KE', 'Kenya'), ('KI', 'Kiribati'), ('KP', "Korea, Democratic People's Republic of"), ('KR', 'Korea, Republic of'), ('KW', 'Kuwait'), ('KG', 'Kyrgyzstan'), ('LA', "Lao People's Democratic Republic"), ('LV', 'Latvia'), ('LB', 'Lebanon'), ('LS', 'Lesotho'), ('LR', 'Liberia'), ('LY', 'Libya'), ('LI', 'Liechtenstein'), ('LT', 'Lithuania'), ('LU', 'Luxembourg'), ('MO', 'Macao'), ('MK', 'Macedonia, the former Yugoslav Republic of'), ('MG', 'Madagascar'), ('MW', 'Malawi'), ('MY', 'Malaysia'), ('MV', 'Maldives'), ('ML', 'Mali'), ('MT', 'Malta'), ('MH', 'Marshall Islands'), ('MQ', 'Martinique'), ('MR', 'Mauritania'), ('MU', 'Mauritius'), ('YT', 'Mayotte'), ('MX', 'Mexico'), ('FM', 'Micronesia, Federated States of'), ('MD', 'Moldova, Republic of'), ('MC', 'Monaco'), ('MN', 'Mongolia'), ('ME', 'Montenegro'), ('MS', 'Montserrat'), ('MA', 'Morocco'), ('MZ', 'Mozambique'), ('MM', 'Myanmar'), ('NA', 'Namibia'), ('NR', 'Nauru'), ('NP', 'Nepal'), ('NL', 'Netherlands'), ('NC', 'New Caledonia'), ('NZ', 'New Zealand'), ('NI', 'Nicaragua'), ('NE', 'Niger'), ('NG', 'Nigeria'), ('NU', 'Niue'), ('NF', 'Norfolk Island'), ('MP', 'Northern Mariana Islands'), ('NO', 'Norway'), ('OM', 'Oman'), ('PK', 'Pakistan'), ('PW', 'Palau'), ('PS', 'Palestinian Territory, Occupied'), ('PA', 'Panama'), ('PG', 'Papua New Guinea'), ('PY', 'Paraguay'), ('PE', 'Peru'), ('PH', 'Philippines'), ('PN', 'Pitcairn'), ('PL', 'Poland'), ('PT', 'Portugal'), ('PR', 'Puerto Rico'), ('QA', 'Qatar'), ('RE', 'Réunion'), ('RO', 'Romania'), ('RU', 'Russian Federation'), ('RW', 'Rwanda'), ('BL', 'Saint Barthélemy'), ('SH', 'Saint Helena, Ascension and Tristan da Cunha'), ('KN', 'Saint Kitts and Nevis'), ('LC', 'Saint Lucia'), ('MF', 'Saint Martin (French part)'), ('PM', 'Saint Pierre and Miquelon'), ('VC', 'Saint Vincent and the Grenadines'), ('WS', 'Samoa'), ('SM', 'San Marino'), ('ST', 'Sao Tome and Principe'), ('SA', 'Saudi Arabia'), ('SN', 'Senegal'), ('RS', 'Serbia'), ('SC', 'Seychelles'), ('SL', 'Sierra Leone'), ('SG', 'Singapore'), ('SX', 'Sint Maarten (Dutch part)'), ('SK', 'Slovakia'), ('SI', 'Slovenia'), ('SB', 'Solomon Islands'), ('SO', 'Somalia'), ('ZA', 'South Africa'), ('GS', 'South Georgia and the South Sandwich Islands'), ('SS', 'South Sudan'), ('ES', 'Spain'), ('LK', 'Sri Lanka'), ('SD', 'Sudan'), ('SR', 'Suriname'), ('SJ', 'Svalbard and Jan Mayen'), ('SZ', 'Swaziland'), ('SE', 'Sweden'), ('CH', 'Switzerland'), ('SY', 'Syrian Arab Republic'), ('TW', 'Taiwan, Province of China'), ('TJ', 'Tajikistan'), ('TZ', 'Tanzania, United Republic of'), ('TH', 'Thailand'), ('TL', 'Timor-Leste'), ('TG', 'Togo'), ('TK', 'Tokelau'), ('TO', 'Tonga'), ('TT', 'Trinidad and Tobago'), ('TN', 'Tunisia'), ('TR', 'Turkey'), ('TM', 'Turkmenistan'), ('TC', 'Turks and Caicos Islands'), ('TV', 'Tuvalu'), ('UG', 'Uganda'), ('UA', 'Ukraine'), ('AE', 'United Arab Emirates'), ('GB', 'United Kingdom'), ('US', 'United States'), ('UM', 'United States Minor Outlying Islands'), ('UY', 'Uruguay'), ('UZ', 'Uzbekistan'), ('VU', 'Vanuatu'), ('VE', 'Venezuela, Bolivarian Republic of'), ('VN', 'Viet Nam'), ('VG', 'Virgin Islands, British'), ('VI', 'Virgin Islands, U.S.'), ('WF', 'Wallis and Futuna'), ('EH', 'Western Sahara'), ('YE', 'Yemen'), ('ZM', 'Zambia'), ('ZW', 'Zimbabwe')]
code2name = {c[0]: c[1] for c in countries}
code2name

{'ww': 'Worldwide',
 'af': 'Africa',
 'an': 'Antarctica',
 'as': 'Asia',
 'eu': 'Europe',
 'na': 'North America',
 'oc': 'Oceania',
 'sa': 'South America',
 'AF': 'Afghanistan',
 'AX': '√Öland Islands',
 'AL': 'Albania',
 'DZ': 'Algeria',
 'AS': 'American Samoa',
 'AD': 'Andorra',
 'AO': 'Angola',
 'AI': 'Anguilla',
 'AQ': 'Antarctica',
 'AG': 'Antigua and Barbuda',
 'AR': 'Argentina',
 'AM': 'Armenia',
 'AW': 'Aruba',
 'AU': 'Australia',
 'AT': 'Austria',
 'AZ': 'Azerbaijan',
 'BS': 'Bahamas',
 'BH': 'Bahrain',
 'BD': 'Bangladesh',
 'BB': 'Barbados',
 'BY': 'Belarus',
 'BE': 'Belgium',
 'BZ': 'Belize',
 'BJ': 'Benin',
 'BM': 'Bermuda',
 'BT': 'Bhutan',
 'BO': 'Bolivia, Plurinational State of',
 'BQ': 'Bonaire, Sint Eustatius and Saba',
 'BA': 'Bosnia and Herzegovina',
 'BW': 'Botswana',
 'BV': 'Bouvet Island',
 'BR': 'Brazil',
 'IO': 'British Indian Ocean Territory',
 'BN': 'Brunei Darussalam',
 'BG': 'Bulgaria',
 'BF': 'Burkina Faso',
 'BI': 'Burundi',
 'KH': 'Cambodia',
 'CM': 'Came

# Validation & Load

In [5]:
def get_df_all_from_files(files):
    df_all = {}
    for file in files:
        path_file = dir_item + file
#         print(path_file)
        is_valid = validate_file(path_file)
        if is_valid:
            hidden_region = get_hidden_region_from_file(file)
            try:
                df_all[hidden_region] = pd.read_csv(path_file)
            except:
                with open(path_file) as f:
                    errored_text = f.read()
                print(f"Error hidden region: {hidden_region}")
                print(errored_text)

    return df_all


df_all = get_df_all_from_files(files)
df_all['ww']

Date  Chrome  Safari  UC Browser  Firefox  Opera    IE  \
0   2018-01   56.31   14.44        8.29     5.66   3.96  3.20   
1   2018-02   57.46   14.39        7.91     5.50   3.69  3.06   
2   2018-03   57.64   14.81        7.37     5.38   3.65  3.11   
3   2018-04   57.36   13.96        7.88     5.45   3.83  3.21   
4   2018-05   58.09   13.78        8.16     5.27   3.68  3.08   
5   2018-06   58.94   13.70        7.46     5.17   3.50  3.12   
6   2018-07   59.69   13.85        7.03     5.02   3.35  3.01   
7   2018-08   59.67   14.51        6.28     4.93   3.50  3.03   
8   2018-09   60.60   14.85        5.42     5.01   3.38  2.85   
9   2018-10   61.55   15.13        4.41     5.02   3.16  2.88   
10  2018-11   61.77   15.09        4.22     4.92   3.15  2.81   
11  2018-12   62.28   14.69        4.21     4.93   3.13  2.55   
12  2019-01   61.72   15.23        3.83     4.66   2.91  2.70   

    Samsung Internet  Edge  Android  ...    BlackBerry  Nokia  Unknown  \
0               2.84  1.89     1.79  ...          0.07   0.09     0.04   
1               2.92  1.86     1.72  ...          0.07   0.05     0.03   
2               2.72  1.93     1.62  ...          0.06   0.04     0.03   
3               2.69  1.98     1.59  ...          0.06   0.05     0.03   
4               2.59  1.91     1.51  ...          0.05   0.04     0.03   
5               2.67  1.89     1.56  ...          0.04   0.04     0.03   
6               2.71  1.88     1.56  ...          0.04   0.04     0.03   
7               2.87  1.90     1.42  ...          0.04   0.04     0.03   
8               2.82  1.99     1.31  ...          0.03   0.03     0.03   
9               2.76  2.08     1.18  ...          0.03   0.02     0.03   
10              2.73  2.15     1.18  ...          0.02   0.03     0.03   
11              2.94  2.03     1.18  ...          0.02   0.03     0.03   
12              3.64  2.13     1.15  ...          0.02   0.03     0.03   

    360 Safe Browser  Vivaldi  Phantom  Mozilla  Tizen  Pale Moon  Other  
0               0.03     0.02     0.03     0.01   0.04       0.02   0.12  
1               0.02     0.02     0.03     0.01   0.02       0.01   0.10  
2               0.02     0.02     0.03     0.01   0.02       0.01   0.10  
3               0.03     0.02     0.03     0.01   0.01       0.01   0.11  
4               0.02     0.02     0.04     0.02   0.01       0.02   0.10  
5               0.02     0.02     0.03     0.02   0.01       0.01   0.10  
6               0.02     0.02     0.03     0.02   0.01       0.01   0.09  
7               0.07     0.02     0.02     0.02   0.01       0.01   0.09  
8               0.03     0.02     0.00     0.02   0.01       0.01   0.08  
9               0.01     0.02     0.00     0.03   0.00       0.01   0.08  
10              0.01     0.02     0.00     0.03   0.00       0.01   0.10  
11              0.02     0.02     0.00     0.02   0.00       0.01   0.08  
12              0.03     0.02     0.00     0.03   0.00       0.01   0.07  

[13 rows x 31 columns]

# Merge df_all

In [6]:
def get_df_merged(df_all):
    df_merged = pd.DataFrame()
    for region_code, df in df_all.items():
        df_tmp = df
        df_tmp['Code'] = region_code
        df_tmp['Region'] = code2name[region_code]
        df_merged = df_merged.append(df_tmp)
    return df_merged

In [7]:
df_merged = get_df_merged(df_all)

/home/jake/.local/lib/python3.6/site-packages/pandas/core/frame.py:6211: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


In [8]:
print(len(df_merged))
df_merged.head(10)

3341


360 Safe Browser  Amigo  Android  Baidu Browser  BlackBerry  Bolt  Camino  \
0              0.01    NaN     1.60           0.03         NaN   NaN     NaN   
1              0.01    NaN     1.47           0.03         NaN   NaN     NaN   
2              0.01    NaN     1.37           0.02         NaN   NaN     NaN   
3              0.01    NaN     1.20           0.02         NaN   NaN     NaN   
4              0.01    NaN     1.07           0.02         NaN   NaN     NaN   
5              0.02    NaN     1.04           0.02         NaN   NaN     NaN   
6              0.01    NaN     1.01           0.02         NaN   NaN     NaN   
7              0.05    NaN     0.99           0.02         NaN   NaN     NaN   
8              0.02    NaN     1.01           0.02         NaN   NaN     NaN   
9              0.01    NaN     0.87           0.02         NaN   NaN     NaN   

   Chrome  Chromium  Coast  ...    Tizen UC Browser  Unknown Unnamed: 1  \
0   73.41      0.01    NaN  ...      NaN       1.34      NaN        NaN   
1   73.97      0.01    NaN  ...      NaN       1.39      NaN        NaN   
2   73.71      0.01    NaN  ...      NaN       1.39      NaN        NaN   
3   73.50      0.01    NaN  ...      NaN       1.22      NaN        NaN   
4   74.91      0.01    NaN  ...      NaN       0.97      NaN        NaN   
5   75.51      0.01    NaN  ...      NaN       0.89      NaN        NaN   
6   76.09      0.01    NaN  ...      NaN       0.89      NaN        NaN   
7   75.73      0.01    NaN  ...      NaN       1.46      NaN        NaN   
8   75.92      0.01    NaN  ...      NaN       1.55      NaN        NaN   
9   76.06      0.01    NaN  ...      NaN       1.24      NaN        NaN   

   Vivaldi  Waterfox  Whale  Whale Browser  Yandex Browser  mCent  
0      NaN       NaN    NaN            NaN             NaN    NaN  
1      NaN       NaN    NaN            NaN             NaN    NaN  
2      NaN       NaN    NaN            NaN             NaN    NaN  
3      NaN       NaN    NaN            NaN             NaN    NaN  
4      NaN       NaN    NaN            NaN             NaN    NaN  
5      NaN       NaN    NaN            NaN             NaN    NaN  
6      NaN       NaN    NaN            NaN             NaN    NaN  
7      NaN       NaN    NaN            NaN             NaN    NaN  
8      NaN       NaN    NaN            NaN             NaN    NaN  
9      NaN       NaN    NaN            NaN             NaN    NaN  

[10 rows x 76 columns]

# Columns

In [9]:
df_merged.columns

Index(['360 Safe Browser', 'Amigo', 'Android', 'Baidu Browser', 'BlackBerry',
       'Bolt', 'Camino', 'Chrome', 'Chromium', 'Coast', 'Coc Coc', 'Code',
       'Comodo Dragon', 'Date', 'DoCoMo', 'Dolfin', 'Edge', 'Firefox', 'Flock',
       'IE', 'IEMobile', 'Iceweasel', 'Iron', 'Jasmine', 'K-Meleon', 'KaiOS',
       'Kindle', 'Lunascape', 'Maxthon', 'MeeGo', 'Microsoft-WebDAV', 'Midori',
       'Minefield', 'Mozilla', 'NetFront', 'NetFront NX', 'Netscape',
       'Nintendo 3DS', 'Nokia', 'Obigo', 'OmniWeb', 'Openwave', 'Opera',
       'Other', 'Pale Moon', 'Phantom', 'Puffin', 'QQ Browser', 'Region',
       'RockMelt', 'Safari', 'Samsung', 'Samsung Internet', 'SeaMonkey',
       'Silk', 'Sleipnir', 'Sogou Explorer', 'Sony PS3', 'Sony PS4',
       'Sony PSP', 'Sony PSP Vita', 'SonyEricsson', 'Swing',
       'Tencent Traveler', 'TheWorld', 'Thunderbird', 'Tizen', 'UC Browser',
       'Unknown', 'Unnamed: 1', 'Vivaldi', 'Waterfox', 'Whale',
       'Whale Browser', 'Yandex Browser', 'mCent

In [10]:
idx_latest = df_merged['Date'] == '2019-01'
cols_major = ['Chrome', 'Safari', 'IE', 'Firefox', 'Edge', 'Samsung Internet', 'Android', 'Opera', 'UC Browser']
cols_region_date = ['Region', 'Date']
cols_major2 = cols_region_date + cols_major
df_major = df_merged.loc[idx_latest, cols_major2]
df_major.head()

Region     Date  Chrome  Safari    IE  Firefox  \
12                           Thailand  2019-01   75.57   13.70  0.86     2.58   
12                            Andorra  2019-01   54.19   29.55  2.76     4.92   
12                         Antarctica  2019-01   64.27   15.38  2.98     5.66   
12  Venezuela, Bolivarian Republic of  2019-01   78.29    2.88  1.85     7.82   
12   Bonaire, Sint Eustatius and Saba  2019-01   51.61   27.83  4.62     3.32   

    Edge  Samsung Internet  Android  Opera  UC Browser  
12  0.64              3.73     0.49   1.26        0.90  
12  3.61              2.89     0.61   1.11        0.01  
12  2.30              6.81     0.23   2.22        0.08  
12  0.54              1.05     2.33   3.96        0.22  
12  3.86              6.98     0.96   0.60        0.01

# IE 순위

In [11]:
df_major.sort_values('IE', ascending=False).head(20)

Region     Date  Chrome  Safari  \
12                 Holy See (Vatican City State)  2019-01   58.53    7.05   
12                                       Bermuda  2019-01   41.60   28.89   
12                                 Liechtenstein  2019-01   36.17   28.85   
12                                √Öland Islands  2019-01   39.56   27.94   
5   Saint Helena, Ascension and Tristan da Cunha  2019-01   76.64    3.54   
12                                        Norway  2019-01   51.97   24.52   
5                                 Western Sahara  2019-01   75.93    1.97   
12                            Korea, Republic of  2019-01   56.04   17.36   
12                                         Japan  2019-01   42.90   33.84   
12                                    Guadeloupe  2019-01   53.91   18.82   
12                     Iran, Islamic Republic of  2019-01   58.85    4.76   
12                                  Cook Islands  2019-01   76.94    9.80   
12                       Virgin Islands, British  2019-01   49.26   20.81   
12                                        Monaco  2019-01   44.53   29.80   
12                                 Faroe Islands  2019-01   33.78   40.65   
5                                       Pitcairn  2019-01   58.33    0.00   
12                                      Botswana  2019-01   61.24    6.44   
12                                   Switzerland  2019-01   36.89   33.31   
12                                      Guernsey  2019-01   37.83   40.64   
12                                       Vanuatu  2019-01   63.54   10.51   

       IE  Firefox  Edge  Samsung Internet  Android  Opera  UC Browser  
12  17.50    10.99  4.57              0.20     0.12   1.01        0.00  
12  15.70     3.27  7.64              1.81     0.65   0.23        0.05  
12  13.98     9.17  6.90              3.51     0.31   0.63        0.00  
12  11.95     4.68  7.56              2.01     0.18   0.82        0.28  
5   11.40     1.77  4.05              0.93     1.58   0.00        0.00  
12  11.24     3.53  3.61              3.05     0.18   1.53        0.03  
5   10.72     1.53  0.00              3.94     5.69   0.22        0.00  
12  10.23     0.61  1.45             12.16     0.29   0.40        0.08  
12   9.45     6.01  4.16              1.01     0.70   0.64        0.25  
12   9.17     6.62  4.17              4.78     0.82   0.84        0.03  
12   9.10    13.39  0.80              8.34     2.15   1.95        0.24  
12   8.84     2.00  0.48              1.45     0.21   0.17        0.05  
12   8.81     4.85  7.17              3.18     1.07   0.69        1.33  
12   8.74     9.97  3.00              1.71     0.13   1.82        0.03  
12   8.52     4.28  6.85              5.36     0.14   0.25         NaN  
5    8.33    33.33   NaN               NaN      NaN   0.00        0.00  
12   8.10     8.35  3.38              3.10     1.56   4.70        1.64  
12   7.68     9.81  5.44              4.79     0.38   1.12        0.06  
12   7.29     4.78  4.78              2.76     1.21   0.26        0.07  
12   7.11    10.48  2.44              2.70     2.08   0.33        0.65

# Export

In [19]:
# df_major[['Region', 'Date', 'IE']].sort_values('IE', ascending=False).head(9).to_csv(dir_item + 'tableau_data/ie_ranking.tsv', sep='\t', index=False)

# 제로데이 이후 점유율 하락?

In [13]:
list_continent = ['ww', 'af', 'an', 'as', 'eu', 'na', 'oc', 'sa',]
list_major_country = ['KR', 'US', 'CN', 'JP']

idx_continent = df_merged['Code'].isin(list_continent)
idx_major_country = df_merged['Code'].isin(list_major_country)
idx_nov_dec = df_merged['Date'].isin(['2018-11', '2018-12'])

df_merged.loc[(idx_continent | idx_major_country) & idx_nov_dec, cols_major2]

Region     Date  Chrome  Safari     IE  Firefox  Edge  \
10          Antarctica  2018-11   54.61   15.16  12.71    10.36  4.80   
11          Antarctica  2018-12   70.59   18.33   4.01     2.90  1.28   
10              Europe  2018-11   59.68   17.42   3.45     8.32  3.24   
11              Europe  2018-12   59.94   17.21   3.19     8.39  3.10   
10               Japan  2018-11   44.28   32.54   9.49     6.31  4.14   
11               Japan  2018-12   44.47   31.96   9.21     6.81  4.00   
10               China  2018-11   49.59   10.76   3.51     2.20  1.61   
11               China  2018-12   50.42   11.16   3.13     2.23  1.63   
10       South America  2018-11   82.56    5.53   1.13     4.34  1.14   
11       South America  2018-12   82.12    5.30   1.18     4.91  1.16   
10                Asia  2018-11   64.52    8.47   1.90     3.00  0.84   
11                Asia  2018-12   65.36    8.57   1.64     3.13  0.79   
10       United States  2018-11   49.13   32.38   5.06     4.84  4.26   
11       United States  2018-12   49.29   31.95   5.00     4.79  4.33   
10       North America  2018-11   52.68   29.47   4.63     4.78  3.98   
11       North America  2018-12   52.47   29.39   4.55     4.79  4.03   
10  Korea, Republic of  2018-11   56.33   14.72  15.98     0.74  1.86   
11  Korea, Republic of  2018-12   58.08   16.04  12.35     1.19  1.66   
10           Worldwide  2018-11   61.77   15.09   2.81     4.92  2.15   
11           Worldwide  2018-12   62.28   14.69   2.55     4.93  2.03   
10             Oceania  2018-11   51.59   32.34   3.46     4.26  3.79   
11             Oceania  2018-12   50.61   32.96   3.05     4.72  3.58   
10              Africa  2018-11   60.68    6.78   1.55     5.62  1.21   
11              Africa  2018-12   61.85    6.31   1.34     5.73  1.07   

    Samsung Internet  Android  Opera  UC Browser  
10              0.19     0.00   1.13        0.00  
11              0.51     0.60   1.45        0.00  
10              2.93     0.95   2.30        0.18  
11              3.13     0.92   2.36        0.22  
10              0.73     0.69   0.54        0.26  
11              0.83     0.74   0.60        0.30  
10              0.42     3.64   0.36       11.54  
11              0.46     3.14   0.34       13.03  
10              1.72     0.64   1.69        0.55  
11              1.69     0.63   1.77        0.51  
10              3.19     1.39   4.03        9.40  
11              3.43     1.36   3.77        8.71  
10              1.85     1.20   0.51        0.18  
11              1.97     1.30   0.55        0.26  
10              1.89     1.10   0.66        0.21  
11              2.02     1.19   0.68        0.27  
10              8.07     0.30   0.41        0.09  
11              8.38     0.31   0.43        0.10  
10              2.73     1.18   3.15        4.22  
11              2.94     1.18   3.13        4.21  
10              2.10     0.52   0.73        0.40  
11              2.39     0.60   0.80        0.44  
10              3.43     2.10  14.76        2.78  
11              3.49     1.99  14.33        2.75

한국 15.98에서 12.35로 떨어짐. 부동층이 많다? 남극 12.71에서 4.01로 폭락. 남아메리카는 오히려 증가. 다른 대륙 및 나라는 조금 떨어짐.

# 네이버 웨일 점유율?

In [14]:
idx_whale = ~df_merged['Whale'].isna()
cols_whale = cols_major2 + ['Whale']

df_merged.loc[idx_whale & idx_latest, cols_whale]

Region     Date  Chrome  Safari     IE  Firefox  Edge  \
12  Korea, Republic of  2019-01   56.04   17.36  10.23     0.61  1.45   

    Samsung Internet  Android  Opera  UC Browser  Whale  
12             12.16     0.29    0.4        0.08   0.07

네이버 웨일 점유율 한국이 유일하게 0.07%. 물론 다른 나라에 사용자가 있겠지만 통계에 잡히지 않았다.

# 주요국 IE 사용 추이

In [15]:
cols_ie = cols_region_date + ['IE']

df_merged.loc[idx_major_country, cols_ie].pivot(index='Region', columns='Date', values='IE')

Date                2018-01  2018-02  2018-03  2018-04  2018-05  2018-06  \
Region                                                                     
China                  3.82     3.04     4.05     4.14     3.05     2.84   
Japan                 16.43    16.33    16.84    16.54    14.09    11.50   
Korea, Republic of    15.81    15.28    16.95    17.22    18.15    15.93   
United States          5.26     4.96     5.24     5.72     5.58     6.01   

Date                2018-07  2018-08  2018-09  2018-10  2018-11  2018-12  \
Region                                                                     
China                  2.56     3.46     4.11     3.76     3.51     3.13   
Japan                 11.13    11.15    10.43    10.20     9.49     9.21   
Korea, Republic of    14.02    13.17    13.45    15.12    15.98    12.35   
United States          5.59     5.48     4.96     5.22     5.06     5.00   

Date                2019-01  
Region                       
China                  2.91  
Japan                  9.45  
Korea, Republic of    10.23  
United States          5.29

5-6월에 2.2% 떨어졌다. 11-12월에 3.63% 떨어졌다. 두 달 모두 '제로데이' 취약점이 발견된 달
5월에 제로데이 취약점이 발견됐는데 10월에 다시 15%까지 회복했다는 사실;;
취약점 발견되기 전인 2018년 5월까지 오히려 증가했다.

# Export

In [21]:
# df_merged.loc[idx_major_country, cols_ie].pivot(index='Region', columns='Date', values='IE').reset_index().to_csv(dir_item + 'tableau_data/line.tsv', sep='\t', index=False)